### PR TITLE
Use Strings for employee IDs to allow IDs with leading zeros

### PR DESCRIPTION
**Pull request by I-Al-Istannen**
## About
This PR changes the employee id field to a String, allowing leading zeros.

## Problems
It might not be backwards compatible with existing global JSONs.

### CODE REVIEW

**Comment by larsk21** [comment]
The parsing library is able to parse string properties in the JSON to int values in our Java class, ignoring leading zeros. So there is really no need to change code here. Changing the property type to string in our Java code would require additional validation checks for the staff ID.

The reason why an exception is thrown when reading integer values with leading zeros, is that in JSON (as well as in other languages), leading zeros indicate octal numbers (which are not supported in JSON). So there are two solutions here:
1. don't use integers with leading zeros (as the resulting octal number does not mean what you intend it to mean)
2. change the property definition of `staffId` in `schemas/global.json` to `{ "type": "string", "pattern": "^[0-9]+$" }`

The question is: are there staff IDs issued with leading zeros?

**Comment by I-Al-Istannen** [comment]
> Edit: If there are staff IDs that include leading zeros by definition, then in order to preserve the leading zeros in the output document, we would have to use a string property.

Yes, that is why I changed it to a String. It will not be parsed to octal *as it is never parsed*. It is dumped out as the user entered it.

> change the property definition of staffId in schemas/global.json to `{ "type": "string", "pattern": "^[0-9]+$" }`

I already changed the first part of that (type: string) but haven't assigned it a pattern yet.

> The question is: are there staff IDs issued with leading zeros?

The bank transfer includes the leading zero. If that is just for show, it probably does not have to be changed in the generator here and that PR can be closed.

**Comment by larsk21** [comment]
Yes, the octal number definition is only relevant if you want to ignore leading zeros in the user input. If the ID does in fact include leading zeros, we would need to use the string representation.

I guess there are ID definitions including leading zeros, but we should check what the allowed values for staff IDs in our case are.

**Comment by larsk21** [comment]
Maybe @gstuer can find out about this?

**Comment by gstuer** [comment]
As far as I know, the KIT only uses numeric values. Therefore, a String is not the way to go. Moreover, leading zeros could also be realized with Integer values. This would solve the same problem without breaking with objekt-orientation principles. However, it may happen that at some point we are going to need non-numerical values at the TimeSheetGenerator. Due to the current situation, the project is focused on stability and reliability. Such a major non-needed change would entail more risks than chances at the moment.

In conclusion, this PR has to wait for implementation until we evaluate the situation and necessity of a change. As far as the current situation allows me to, I am going to gather information by talking with the administrative person in charge at KIT. Stay tuned.

**Comment by larsk21** [comment]
Just as a side note: realizing leading zeros with integers would require us to know the length of the ID in order to write the ID in the output document.

**Comment by dfuchss** [comment]
Closing for now. If the PR shall be reopened, just do that :)
